### PR TITLE
Unification CI Presubmit and Nightly: Step 2

### PR DESCRIPTION
**Pull request by DwarKapex**
**This is the second step towards unification of the nightly and presubmit workflows.**

What is done:
1. Bump the manifest file in-place with the base container build, when the workflow is scheduled as a nightly run.
2. Remove the trial branch because it is no longer needed: the bumped manifest will be stored inside the base container.
3. In `_ci.yaml`, save the tags for all generated Docker images (both meal-kit and final) in a single JSON object (to get rid of long and hard-to-read variables in the `outputs` section) to be able to publish them in the `ci.yaml` workflow (`publish` job).
4. Move single-use workflows (`_test_distribution.yaml`) into the `ci.yaml` workflow to prevent hitting the 20-workflow limit for the CI.

**The next step is to add another job to update the manifest in the main branch**:
A. If the pipeline is all green, then create a PR with a bumped manifest, approve it, and merge it automatically.
B. If the pipeline is not all green, then create a PR and ping the maintainer of the repo (@yhtang, @terrykong) to perform a manual check of the pipeline and merge the PR manually if it’s acceptable.

### CODE REVIEW

File: .github/workflows/ci.yaml
Comment (terrykong) [inline .github/workflows/ci.yaml:44]: You are missing `on.workflow_dispatch.inputs.PUBLISH`

File: .github/workflows/ci.yaml
Comment (terrykong) [inline .github/workflows/ci.yaml:65]: `github.event_name == 'schedule' ` looks like it's already captured in `needs.metadata.outputs.PUBLISH`

File: .github/workflows/ci.yaml
Comment (terrykong) [inline .github/workflows/ci.yaml:67]: I can't think of a case where one of these fails and the others don't (maybe an API failure?). But what do you think about letting it be `false` so that we at least get some images published

File: .github/workflows/ci.yaml
Comment (terrykong) [inline .github/workflows/ci.yaml:93]: Same comment about `github.event_name == 'schedule'` being included in `needs.metadata.outputs.PUBLISH`

File: .github/workflows/ci.yaml
Comment (terrykong) [inline .github/workflows/ci.yaml:124]: `needs.metadata.outputs.PUBLISH`?

File: .github/workflows/ci.yaml
Comment (terrykong) [inline .github/workflows/ci.yaml:80]: I'm not sure casing matters, but should this be `t5x`

**Review by terrykong — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yaml
Comment (yhtang) [reply .github/workflows/ci.yaml:80]: Tags are indeed case-sensitive.

File: .github/workflows/_build_base.yaml
Comment (yhtang) [inline .github/workflows/_build_base.yaml:99]: For better readability, I recommend the pattern where a dedicated predicate step would run the logic to determine whether version bump should happen, and then store the outcome as its step output. The actual bump step will then be conditional on the output of the predicate step. Take for example how the `PUBLISH` output is set in the `metadata` step of many nightly workflows.

Comment (yhtang) [reply .github/workflows/_build_base.yaml:99]: This way, it will also be easier to visually examine if the bump step did run in the GitHub Web GUI.

File: .github/workflows/_build_jax.yaml
Comment (yhtang) [inline .github/workflows/_build_jax.yaml:40]: Is this duplicating part of #465?

File: .github/workflows/_ci.yaml
Comment (yhtang) [inline .github/workflows/_ci.yaml:18]: Maybe we should just use a single JSON object to store all the mealkit/final tags, respectively. This is just a thought and please feel free to throw in yours :)

File: .github/workflows/ci.yaml
Comment (yhtang) [inline .github/workflows/ci.yaml:14]: Does the `workflow_dispatch` section still need the `TRIAL_BRANCH` input?

File: .github/workflows/ci.yaml
Comment (yhtang) [reply .github/workflows/ci.yaml:44]: Is this already resolved?

File: .github/workflows/ci.yaml
Comment (yhtang) [reply .github/workflows/ci.yaml:67]: +1 for `fail-fast: false`

File: .github/workflows/ci.yaml
Comment (yhtang) [reply .github/workflows/ci.yaml:80]: This is the `image` part of the complete tag, which is required to be in lower case:
```
Error parsing reference: "Fedora:37" is not a valid repository/tag: invalid reference format: repository name must be lowercase
```

**Review by yhtang — CHANGES_REQUESTED** [review]

File: .github/workflows/_ci.yaml
Comment (terrykong) [inline .github/workflows/_ci.yaml:15]: can't comment on the line, but looks like you don't use TRIAL_BRANCH on line 10

File: .github/workflows/ci.yaml
Comment (terrykong) [reply .github/workflows/ci.yaml:14]: If TRIAL_BRANCH isn't used, then lines 52,60 aren't needed

File: .github/workflows/ci.yaml
Comment (terrykong) [reply .github/workflows/ci.yaml:44]: It doesn't look like it to me. The only workflow_dispatch input is TRIAL_BRANCH, but I don't see PUBLISH

File: .github/workflows/_build_base.yaml
Comment (DwarKapex) [reply .github/workflows/_build_base.yaml:99]: Done

File: .github/workflows/_build_jax.yaml
Comment (DwarKapex) [reply .github/workflows/_build_jax.yaml:40]: Yes, because I started this PR before #465

File: .github/workflows/_ci.yaml
Comment (DwarKapex) [reply .github/workflows/_ci.yaml:15]: Removed

File: .github/workflows/_ci.yaml
Comment (DwarKapex) [reply .github/workflows/_ci.yaml:18]: Never used it before, but let me give a shot

Comment (DwarKapex) [reply .github/workflows/_ci.yaml:18]: Not sure it looks better now, but more flexible I think

File: .github/workflows/ci.yaml
Comment (DwarKapex) [reply .github/workflows/ci.yaml:14]: Removed

File: .github/workflows/ci.yaml
Comment (DwarKapex) [reply .github/workflows/ci.yaml:44]: Added PUBLISH to inputs in `workflow_dispatch`

**Comment by DwarKapex** [comment]
Addressed comments above but one.

File: .github/workflows/_ci.yaml
Comment (terrykong) [inline .github/workflows/_ci.yaml:46]: Is `TAG_BASE` used somewhere?

**Comment by yhtang** [comment]
Converting the PR to draft to prevent accidental merge, as it appears to be in active development.

File: .github/workflows/_ci.yaml
Comment (DwarKapex) [reply .github/workflows/_ci.yaml]: Removed

**Comment by yhtang** [comment]
These workflows will be broken if we remove `_finalize.yaml`:
```
$ grep _finalize *.yaml
ci.yaml:    uses: ./.github/workflows/_finalize.yaml
nightly-base-build.yaml:    uses: ./.github/workflows/_finalize.yaml
nightly-jax-build.yaml:    uses: ./.github/workflows/_finalize.yaml
nightly-jax-test-unit.yaml:    uses: ./.github/workflows/_finalize.yaml
nightly-pallas-build.yaml:    uses: ./.github/workflows/_finalize.yaml
nightly-pallas-test-unit.yaml:    uses: ./.github/workflows/_finalize.yaml
nightly-pax-build.yaml:    uses: ./.github/workflows/_finalize.yaml
nightly-pax-test-mgmn.yaml:    uses: ./.github/workflows/_finalize.yaml
nightly-rosetta-pax-build.yaml:    uses: ./.github/workflows/_finalize.yaml
nightly-rosetta-t5x-build-test.yaml:    uses: ./.github/workflows/_finalize.yaml
nightly-t5x-build.yaml:    uses: ./.github/workflows/_finalize.yaml
nightly-t5x-test-mgmn.yaml:    uses: ./.github/workflows/_finalize.yaml
```

In the longer run, all these nightly workflows will no longer exist once we unify the presubmit and the nightlies. However, I don't think we can afford to have them broken during the transition period.

File: .github/workflows/_build.yaml
Comment (yhtang) [inline .github/workflows/_build.yaml:140]: Given that this is a generic and reusable workflow, its output should also have generic names.
The outputs from different calls to this workflow can be distinguished by their job name.
i.e. we want
```
${{ needs.build_A.outputs.TAG_MEALKIT }}
${{ needs.build_B.outputs.TAG_MEALKIT }}
```
instead of
```
${{ needs.build_A.outputs.TAG_A_MEALKIT }}
${{ needs.build_B.outputs.TAG_B_MEALKIT }}
```

File: .github/workflows/_build_base.yaml
Comment (yhtang) [inline .github/workflows/_build_base.yaml:55]: This logic should be placed directly in `ci.yaml` since it is a global condition. `_build_base.yaml` can take the output of the logic as an input.

File: .github/workflows/_build_rosetta.yaml
Comment (yhtang) [inline .github/workflows/_build_rosetta.yaml:147]: Similar comment like those for `_build.yaml`.

File: .github/workflows/_ci.yaml
Comment (yhtang) [inline .github/workflows/_ci.yaml:11]: Since this is the only output related to tags, `ALL_` won't be necessary. How about just `CONTAINER_TAGS`?

File: .github/workflows/_ci.yaml
Comment (yhtang) [inline .github/workflows/_ci.yaml:12]: Maybe clarify these are docker tags rather than git tags?

File: .github/workflows/_ci.yaml
Comment (yhtang) [inline .github/workflows/_ci.yaml:125]: Need to update the output names once `_build.yaml`/`_build_rosetta.yaml` is updated.

File: .github/workflows/_ci.yaml
Comment (yhtang) [inline .github/workflows/_ci.yaml:131]: Is this consistent with `_build.yaml`:
```
echo "TAG_${CONTAINER_TAG_NAME}_MEALKIT=${{ steps.mealkit-metadata.outputs.tags }}" | tee -a $GITHUB_OUTPUT
```
should it be
```
"TAG_PALLAS_MEALKIT": "${{ needs.build-pallas.outputs.TAG_PALLAS_MEALKIT}}"
```
?

File: .github/workflows/ci.yaml
Comment (yhtang) [inline .github/workflows/ci.yaml:68]: Looks like the correspondence between TAG and TARGET_NAME can be programmed:
- lower case -> upper case
- middle hyphen -> underscore

File: .github/workflows/ci.yaml
Comment (yhtang) [inline .github/workflows/ci.yaml:85]: We probably don't want to publish N images in `ghcr.io/nvidia`, but this is worth discussing with everyone and the product team. @sbhavani

File: .github/workflows/ci.yaml
Comment (yhtang) [inline .github/workflows/ci.yaml:90]: This publication workflow should be turned into a reusable one now that we are repeating it twice.

File: .github/workflows/ci.yaml
Comment (yhtang) [inline .github/workflows/ci.yaml:120]: Instead of breaking encapsulation and inlining the workflow here, how about we try to merge a few `_test_*.yaml` workflows?

**Review by yhtang — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yaml
Comment (terrykong) [inline .github/workflows/ci.yaml:68]: Can this be simplified and derived from ALL_TAGS?

**Comment by terrykong** [comment]
Also, should these lines be removed? https://github.com/NVIDIA/JAX-Toolbox/blob/main/.github/workflows/nightly-base-build.yaml#L5-L6

It looks like both the old nightlies and this new nightly both run at the same time.

File: .github/workflows/ci.yaml
Comment (terrykong) [reply .github/workflows/ci.yaml:85]: Just clarifying, is your suggestion to publish fewer, or push them elsewhere?

File: .github/workflows/_build.yaml
Comment (DwarKapex) [reply .github/workflows/_build.yaml:140]: This is leftover from my experiments. This is step is not required any more. All tags will be saved inside `_ci.yaml` workflow under `save-tags` job

File: .github/workflows/_build_base.yaml
Comment (DwarKapex) [reply .github/workflows/_build_base.yaml:55]: Moved to `ci.yaml`, as well as `BUILD_DATE`

File: .github/workflows/_build_rosetta.yaml
Comment (DwarKapex) [reply .github/workflows/_build_rosetta.yaml:147]: Removed completely

File: .github/workflows/_ci.yaml
Comment (DwarKapex) [reply .github/workflows/_ci.yaml:12]: Done

File: .github/workflows/_ci.yaml
Comment (DwarKapex) [reply .github/workflows/_ci.yaml:11]: Done

File: .github/workflows/_ci.yaml
Comment (DwarKapex) [reply .github/workflows/_ci.yaml:131]: Yes. See comments above.

File: .github/workflows/ci.yaml
Comment (DwarKapex) [reply .github/workflows/ci.yaml:68]: Yeah, I wanted to do that in place, but looks like I need to do it in 2 steps. Or is there anternative?

File: .github/workflows/ci.yaml
Comment (DwarKapex) [reply .github/workflows/ci.yaml:85]: OK, that is something new. As of now, when we run nightly we publish both mealkit and final. Why do we want to change the publishing politics now?

File: .github/workflows/ci.yaml
Comment (DwarKapex) [reply .github/workflows/ci.yaml:90]: Convert it to matrix to avoid introducing another workflow

File: .github/workflows/ci.yaml
Comment (DwarKapex) [reply .github/workflows/ci.yaml:120]: Reverted. Will do some investigation what can be merged to reduce the number of workflows.

File: .github/workflows/ci.yaml
Comment (DwarKapex) [reply .github/workflows/ci.yaml:68]: I tried to prototype it and almost found the solution, but the problem is that I have 2 different `ALL_TAGS` (one from amd64 build, the other one is from arm64 one). I need to combine these 2 `ALL_TAGS` into 1 JSON object somehow  and feed `matrix` with it.
But let's create an issue for further pipeline improvements after I merge this PR.
Agree? @terrykong

File: .github/workflows/ci.yaml
Comment (yhtang) [reply .github/workflows/ci.yaml]: Great that's a better idea.

File: .github/workflows/ci.yaml
Comment (yhtang) [reply .github/workflows/ci.yaml]: If we do it as a job step, then I guess it's not a problem if we use 2 or even more lines to compute it.

File: .github/workflows/ci.yaml
Comment (yhtang) [inline .github/workflows/ci.yaml:76]: `${{...}}` is not required but OK in a `if:` key (this is a GitHub Actions peculiarity IMHO)
Either way, it looks a bit strange since we have half of the condition inside of `${{...}}` and the other half out of it.

File: .github/workflows/ci.yaml
Comment (yhtang) [inline .github/workflows/ci.yaml:94]: Could we use a more descriptive suffix rather than `..._1` and `..._2`?

**Review by yhtang — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yaml
Comment (yhtang) [reply .github/workflows/ci.yaml]: Seems like this is already resolved in the new 'publish' job.

File: .github/workflows/ci.yaml
Comment (yhtang) [inline .github/workflows/ci.yaml:59]: This change is not necessary. Do you have a particular reason?

File: .github/workflows/ci.yaml
Comment (DwarKapex) [reply .github/workflows/ci.yaml]: Ah, got it. Thanks for clarification

File: .github/workflows/ci.yaml
Comment (DwarKapex) [reply .github/workflows/ci.yaml:76]: Removed ${{}} around canceled()

File: .github/workflows/ci.yaml
Comment (DwarKapex) [reply .github/workflows/ci.yaml:59]: Reverted

**Review by terrykong — DISMISSED** [review]
LGTM, but would wait for @yhtang 's approval

**Comment by terrykong** [comment]
Just a note on this comment in the description:
> B. If pipeline is not all green, then create PR and ping developers (Yu-Hang, Terry, me, someone else) to perform a manual check of pipeline and merge the PR manually.

One thing I'm worried about is the PR can become outdated, so it should either be closed or labelled as stale since a later nightly may update the manifest and then this PR would be considered a regression if accidentally merged. Not something to be addressed in this PR, but just a note.

**Comment by DwarKapex** [comment]
> Just a note on this comment in the description:
> 
> > B. If pipeline is not all green, then create PR and ping developers (Yu-Hang, Terry, me, someone else) to perform a manual check of pipeline and merge the PR manually.
> 
> One thing I'm worried about is the PR can become outdated, so it should either be closed or labelled as stale since a later nightly may update the manifest and then this PR would be considered a regression if accidentally merged. Not something to be addressed in this PR, but just a note.

Agree, we need to set sort of timeout for automatically generated PRs.

File: .github/workflows/ci.yaml
Comment (yhtang) [inline .github/workflows/ci.yaml:52]: Please add a workflow input to allow  manual override of `BUMP_MANIFEST`, in case if we need to manually run the nightly workflow.

**Review by yhtang — CHANGES_REQUESTED** [review]
LGTM but please add the `BUMP_MANIFEST` input to `ci.yaml` as commented below 😄 
Also, please revise the PR description for readability and typo, as it will become part of the commit message.

File: .github/workflows/ci.yaml
Comment (DwarKapex) [reply .github/workflows/ci.yaml]: Done

**Comment by yhtang** [comment]
I'm seeing the error when running manually: https://github.com/NVIDIA/JAX-Toolbox/actions/runs/7626810302/job/20774071918#step:8:23

**Comment by DwarKapex** [comment]
> https://github.com/NVIDIA/JAX-Toolbox/actions/runs/7626810302/job/20774071918#step:8:23

Fixed: https://github.com/NVIDIA/JAX-Toolbox/actions/runs/7632038858/job/20791396138#step:7:56
@yhtang

File: .github/workflows/ci.yaml
Comment (terrykong) [inline .github/workflows/ci.yaml:103]: nit:
```suggestion
          echo "TARGET_TAGS=${JSON}" | tee -a $GITHUB_OUTPUT    
```

Comment (DwarKapex) [reply .github/workflows/ci.yaml:103]: Btw, why? What's the difference?

Comment (terrykong) [reply .github/workflows/ci.yaml]: In the orig, you print to stdout and right below you redirect (And append) to GITHUB_OUTPUT.

The suggestion was just to use `tee -a`, which appends, but also passes thru to stdout, so it is effectively the two lines you originally had.

**Review by yhtang — DISMISSED** [review]

**Comment by yhtang** [comment]
Note that after this is merged, we will have two sets of nightly runs: the old ones and the new one. We will continue using the old ones while we test the new one and make sure it is fully functional, before we flip the switch.

**Review by terrykong — CHANGES_REQUESTED** [review]
>Note that after this is merged, we will have two sets of nightly runs: the old ones and the new one. We will continue using the old ones while we test the new one and make sure it is fully functional, before we flip the switch.

Will there be any issues if we let both of them run? One concern I have is both start at the same time, so if someone is looking at the `ghcr.io/nvidia/jax:{nightly,mealkit}...` tags, they won't know which pipeline it is from without digging into the digests

**Comment by yhtang** [comment]
> > Note that after this is merged, we will have two sets of nightly runs: the old ones and the new one. We will continue using the old ones while we test the new one and make sure it is fully functional, before we flip the switch.
> 
> Will there be any issues if we let both of them run? One concern I have is both start at the same time, so if someone is looking at the `ghcr.io/nvidia/jax:{nightly,mealkit}...` tags, they won't know which pipeline it is from without digging into the digests

I've add `if: false` to the publish section of the new `ci.yaml`, so that we won't be publishing any containers using the unified nightly for the time being.

**Review by terrykong — APPROVED** [review]

**Pull request merged**
